### PR TITLE
Boxes: Query algod Box by application ID and Box name

### PR DESCRIPTION
**Pull request by michaeldiamant**
Adds support to query algod for Box by application ID and Box name with unit + integration algorand-sdk-testing support.

Notes:
* Thanks to @algoidurovic for providing a base in #322.
* The PR makes no attempt to provide convenience methods for supplying an encoded box name.  Convenience methods will be provided in a future PR.

### CODE REVIEW

File: test/docker/run_docker.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ set -e
 # reset test harness
 rm -rf test-harness
 rm -rf test/features
-git clone --single-branch --branch feature/box-storage https://github.com/algorand/algorand-sdk-testing.git test-harness
+git clone --single-branch --branch use_latest https://github.com/algorand/algorand-sdk-testing.git test-harness
```
Comment (michaeldiamant) [inline test/docker/run_docker.sh:8]: Reviewer note:  The algorand-sdk-testing branch change can be reverted once https://github.com/algorand/algorand-sdk-testing/pull/204 is merged.

Comment (michaeldiamant) [inline test/docker/run_docker.sh:8]: https://github.com/algorand/go-algorand/pull/4154 is merged.  Safe to revert branch change.

```suggestion
git clone --single-branch --branch feature/box-storage https://github.com/algorand/algorand-sdk-testing.git test-harness
```

File: test/algodclientv2_test.go [hunk 2]
```diff
@@ -256,3 +256,15 @@ func weMakeAnAccountApplicationInformationCallAgainstAccountApplicationID(accoun
...
 	return nil
 }
+
+func weMakeAGetApplicationBoxByNameCallForApplicationIDWithEncodedBoxName(appId int, encodedBoxName string) error {
+	algodClient, err := algod.MakeClient(mockServer.URL, "")
+	if err != nil {
+		return err
+	}
+
+	c := algodClient.GetApplicationBoxByName(uint64(appId))
```
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply test/algodclientv2_test.go]: @algoidurovic Makes sense - I took in the feedback via https://github.com/algorand/go-algorand-sdk/pull/338/commits/9230c4215ad25f3fbd58ee0bcf0e8150fc7c96ba.

Resolving - Feel welcomed to reopen as needed.

File: test/applications_integration_test.go [hunk 3]
```diff
@@ -914,6 +914,26 @@ func checkRandomElementResult(resultIndex int, input string) error {
 	return nil
 }
 
+func theContentsOfTheBoxWithNameShouldBeIfThereIsAnErrorItIs(encodedBoxName, boxContents, errStr string) error {
+
+	c := algodV2client.GetApplicationBoxByName(applicationId)
```
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply test/applications_integration_test.go]: @algoidurovic Makes sense - I took in the feedback via https://github.com/algorand/go-algorand-sdk/pull/338/commits/9230c4215ad25f3fbd58ee0bcf0e8150fc7c96ba.

Resolving - Feel welcomed to reopen as needed.

File: client/v2/algod/getApplicationBoxByName.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/algorand/go-algorand-sdk/client/v2/common/models"
+)
+
+// GetApplicationBoxByNameParams contains all of the query parameters for url serialization.
+type GetApplicationBoxByNameParams struct {
+
+	// Name a box name, in the goal app call arg form 'encoding:value'. For ints, use
+	// the form 'int:1234'. For raw bytes, use the form 'b64:A=='. For printable
+	// strings, use the form 'str:hello'. For addresses, use the form 'addr:XYZ...'.
+	Name string `url:"name,omitempty"`
```
Comment (algochoi) [inline client/v2/algod/getApplicationBoxByName.go:17]: is `omitempty` necessary if this is a mandatory parameter?

**Review by algochoi — APPROVED** [review]
Looks good! Thanks for the changes. One small question, but looks fine to me

File: client/v2/algod/getApplicationBoxByName.go (hunk 4 above)
Comment (michaeldiamant) [reply client/v2/algod/getApplicationBoxByName.go:17]: @algochoi It's a fair question - let me know if my response helps.  Happy to discuss live too.

My 2c:
* There might also be a question if `omitemty` is _ever_ necessary.  Certainly, here it seems a bit strange.
* For the use case at-hand:  It's generated code.  And my unverified guess is _generator_ does not disambiguate required vs optional query string parameters.

**Pull request merged**